racket/src/ChezScheme/build.zuo

```diff
@@ -139,18 +139,28 @@
                          vars))
            token))
 
+  (define doc-project-names
+    ;; happily, the name of the subdirectory is the same as
+    ;; value assigned to `x` in the corresponding makefile
+    '("csug" "release_notes"))
+  (define doc-install-dir-keys
+    ;; in parallel to doc-project-names
+    '(InstallCSUG InstallReleaseNotes))
   (define (build-docs token target)
     (check-not-kernel-only)
-    (define (do-doc name)
+    (define (do-doc name install-dir-key)
       (mkdir-p (at-dir name))
       (define mf (~a "m=" m "\n"
                      "srcdir=" (find-relative-path (at-dir name) (at-source name)) "\n"
+                     ;; FIXME: fales if configured path includes "#"
+                     "installdir=" (hash-ref config install-dir-key) "\n"
                      (file->string (at-source (~a "makefiles/Makefile-" name ".in")))))
       (display-to-file mf (at-dir name "Makefile") :truncate)
       (shell/wait "make" target
                   (hash 'dir (at-dir name))))
```

Comment:
@mflatt I've pushed a rebased branch with the start of the Chez configure changes from https://github.com/racket/racket/pull/4179#issuecomment-1104231034. I've dropped https://github.com/LiberalArtist/racket/commit/483239d9b5c16506b23b1ae0d62dd21841062d1a, at least for now, since it no longer directly applies, but https://github.com/LiberalArtist/racket/releases/tag/pr-4203-v0 should preserve it from garbage collection.

I've used names consistent with the existing Chez `configure` flags, rather than the Autoconf conventions, which would have used e.g. `--docdir=` rather than `--installdoc=`.

There are at few things I'm wondering about:

 0. Is there a mechanism to add quoting around a path destined to be written into a makefile? I tried configuring with a path like `/foo/#bar/baz`, but `make` interpreted the `#` as beginning a comment, so the files were installed to `/foo/`. (I sometimes use paths like this because they are covered by the `.gitignore` file.)
 0. I think there probably should be a `.gitignore` rule for `main.zuo` files generated by `configure`, but I don't have a great sense of how to best to structure one or where to put it. I could imagine putting a directory-specific `.gitignore` file in each directory where a `main.zuo` could be generated or adding the full paths to the top-level `.gitignore`, or maybe it would work to add a general `main.zuo` rule at the top level and put a comment in the placeholder `racket/src/zuo/main.zuo` that it has to be force-added.
 0. I noticed that Zuo is invoked with `	$(ZUO) $(workarea) install-docs MAKE="$(MAKE)"`, but the sub-build is invoked using `"make"`.
 0. I haven't tried building with Guix yet, but I'm suspecting that it will be a problem that this doesn't seem to pass general make flags on to the document `make` invokation: we pass `STEXLIB=`, because we've removed the bundled `stex` directory and replaced it with one we have elsewhere. I could support `STEXLIB` specially via `configure`, but is there a mechanism for propagating flags to `make` more generally?

Reply:
Thanks for these improvements!

> Is there a mechanism to add quoting around a path destined to be written into a makefile?

You can use shell quoting, but escaped by one level if you're calling `configure` in a shell. So, something like `--installdoc="/foo/\"#\"bar/baz"`. Zuo scripts specifically parse information from `Mf-config`/`Makefile` as shell sequences (or, at least, that's how it's meant to work).

> I think there probably should be a .gitignore rule for main.zuo files generated by configure

I have so far resisted trying to exclude more build files. My model in recent times is that you either use a build directory or you're opting into a context where you don't care about mingling build and source files. But probably I can be convinced.

> I noticed that Zuo is invoked with $(ZUO) $(workarea) install-docs MAKE="$(MAKE)", but the sub-build is invoked using "make".

I agree that the Zuo scripts need to be fixed to not use "make".

> I haven't tried building with Guix yet, but I'm suspecting that it will be a problem that this doesn't seem to pass general make flags on to the document make invokation

Based on earlier experiments, I think GNU Make will set environment variables so that `make` arguments get passed along, mostly. And this might be a place where it makes sense to assume/require GNU make.

Reply:
> > I haven't tried building with Guix yet, but I'm suspecting that it will be a problem that this doesn't seem to pass general make flags on to the document make invokation
> 
> Based on earlier experiments, I think GNU Make will set environment variables so that `make` arguments get passed along, mostly. And this might be a place where it makes sense to assume/require GNU make.

Delightfully, this indeed worked! At least for Guix, assuming GNU make is fine.

Reply:
> > Is there a mechanism to add quoting around a path destined to be written into a makefile?
> 
> You can use shell quoting, but escaped by one level if you're calling `configure` in a shell. So, something like `--installdoc="/foo/\"#\"bar/baz"`. Zuo scripts specifically parse information from `Mf-config`/`Makefile` as shell sequences (or, at least, that's how it's meant to work).
> 

Actually, that makes installation fail altogether.

Here are two scenarios:

 1. With `$ ../configure "--installprefix=/tmp/chez-zuo/#sub"`

    in `ta6le/Mf-config`:
    ```
    InstallBin=/tmp/chez-zuo/#sub/bin
    InstallLib=/tmp/chez-zuo/#sub/lib
    InstallMan=/tmp/chez-zuo/#sub/share/man/man1
    InstallCSUG=/tmp/chez-zuo/#sub/share/doc/csug9.5
    InstallReleaseNotes=/tmp/chez-zuo/#sub/share/doc/csv9
    ```
    in `ta6le/csug/Makefile`:
    ```
    m=ta6le
    srcdir=../../../csug
    installdir=/tmp/chez-zuo/#sub/share/doc/csug9.5
    # srcdir, installdir, and m are added before
    ```
    Running `make install-doc` installs to `/tmp/chez-zuo/`, without the subdirectories.
 3. With `../configure "--installprefix=/tmp/chez-zuo/\"#\"sub"`

    in `ta6le/Mf-config`:
    ```
    InstallBin=/tmp/chez-zuo/"#"sub/bin
    InstallLib=/tmp/chez-zuo/"#"sub/lib
    InstallMan=/tmp/chez-zuo/"#"sub/share/man/man1
    InstallCSUG=/tmp/chez-zuo/"#"sub/share/doc/csug9.5
    InstallReleaseNotes=/tmp/chez-zuo/"#"sub/share/doc/csv9
    ```
    in `ta6le/csug/Makefile`:
    ```
    m=ta6le
    srcdir=../../../csug
    installdir=/tmp/chez-zuo/"#"sub/share/doc/csug9.5
    # srcdir, installdir, and m are added before
    ```
    Running `make install-doc` 
    ```
    make[2]: Leaving directory '/home/philip/code/tmp/racket/racket/src/ChezScheme/tmp/ta6le/csug/math/csug'
    touch csug.mathrun
    chmod 644 csug.pdf
    ../../../csug/../makefiles/installsh -m 2755 -d /tmp/chez-zuo/"
    bash: -c: line 1: unexpected EOF while looking for matching `"'
    bash: -c: line 2: syntax error: unexpected end of file
    make[1]: *** [Makefile:21: install] Error 2
    make[1]: Leaving directory '/home/philip/code/tmp/racket/racket/src/ChezScheme/tmp/ta6le/csug'
    shell command failed
     in map
     in build-one
     in loop
     in module->hash
    make: *** [Makefile:85: install-docs] Error 1
    ```

Reply:
On the special characters issue, [the Autoconf manual](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.71/html_node/Special-Chars-in-Variables.html) lists several restrictions on the values of output variables, including a ban on `#`, and says that the restrictions "apply both to the values that `configure` computes, and to the values set directly by the user."

Reply:
I didn't think through the fact that `make` parsing would matter in this case, but it also would not have occurred to me that `make` doesn't follow `sh` quoting rules. Remarkably enough, every Unix `make` variant I've tried works the same way: `#` anywhere ends the line unless it's preceded by `\`, and `\#` escapes the `#`. Naturally, `nmake` on Windows is different.

Probably Zuo's `config-file->hash` should follow the Unix `make` rule for `#` and `\#`, and the small `nmake` mismatch seems unlikely to create new issues. Then, it would be a matter of writing a `configure` invocation that uses a shell encoding of a makefile encoding of a shell encoding, which is possibly `../configure "--installprefix=/tmp/chez-zuo/\"\\#\"sub"`.

Reply:
That makes sense to me, but would the Zuo code writing to makefiles then also need to re-encode `#`? Alternatively, I was surprised by it (given their high tolerance for painfully weird configurations), but I personally could live with [Autoconf's rule](https://www.gnu.org/savannah-checkouts/gnu/autoconf/manual/autoconf-2.71/html_node/Special-Chars-in-Variables.html) against using such characters in arguments to `configure`. Dealing with arcane quoting rules is easily my least-favorite part of traditional build scripting.

Reply:
Yes, and this is a reason to avoid writing makefiles from Zuo scripts. For example, the Racket CS Zuo scripts do not try to write a Chez Scheme configuration "makefiles", and instead they pass variables to other Zuo scripts in Chez Scheme. There is some makefile writing in the Chez Scheme Zuo scripts, but only a little, and probably that could be improved. Anyway, I can sympathize with the autoconf "don't do that" approach.

Reply:
That makes sense, and I think I can avoid doing so here, too.

Tangentially, one format that avoids this is [`environment.d` configuration files](https://www.freedesktop.org/software/systemd/man/environment.d.html#Configuration%20Format) as used by `systemd`. They don't support any kind of quoting at all, and the don't support newlines in variable values, but they do support `$x`, `${x}`, `${a:-b}`, `${a:+b}`, line comments, and blank lines.